### PR TITLE
Keep a fullscreen window in the layout

### PR DESCRIPTION
**Pull request by jwijenbergh**
This is the most simple fix to #1482 I could come up with

Instead of removing a fullscreen window from the layout it keeps it both in the floating layout and tiled layout. This needs group to keep track of the windows that are in the tiled layouts as these layouts do not have a way to get the clients in the layout (`.clients` does not work for all layouts)

Note that this does not fix the behaviour that when a floating window is set to fullscreen and then non fullscreen again it will be tiled again.

### CODE REVIEW

**Comment by jwijenbergh** [comment]
Steps taken:

- Rebased with master
- Added tests for the tiled windows list and re-ordering of fullscreen windows that is now fixed
- Marked ready for review

File: libqtile/group.py [hunk 1]
```diff
@@ -285,20 +292,29 @@ def mark_floating(self, win, floating):
...
         else:
             self.floating_layout.remove(win)
             self.floating_layout.blur()
             for i in self.layouts:
-                i.add(win)
+                # A window that was fullscreen should only be added if it was not a tiled window
+                if win not in self.tiled_windows:
+                    i.add(win)
                 if win is self.current_window:
                     i.focus(win)
```
Comment (m-col) [inline libqtile/group.py:315]: We can save more checks here too (especially as `win not in self.tiled_windows` is O(n)):
```suggestion
            # A window that was fullscreen should only be added if it was not a tiled window
            if win not in self.tiled_windows:
		for i in self.layouts:
                    i.add(win)
	    if win is self.current_window:
		for i in self.layouts:
                    i.focus(win)
```

File: libqtile/group.py [hunk 2]
```diff
@@ -46,6 +46,7 @@ def __init__(self, name, layout=None, label=None):
         self.label = name if label is None else label
         self.custom_layout = layout  # will be set on _configure
         self.windows = []
+        self.tiled_windows = []
```
Comment (m-col) [inline libqtile/group.py:49]: Looks like we don't need to preserve order of this list, so let's make it a `set` so that duplicate items are impossible.

**Review by m-col — CHANGES_REQUESTED** [review]
Overall this looks great, thanks for this! I've added some review comments, just some tiny changes.

Could you also add a note to the changelog? We used to keep a bugfixes section there but I see now it hasn't been included since 0.16.0! Might be good to keep that up.

File: libqtile/group.py [hunk 3]
```diff
@@ -285,20 +292,29 @@ def mark_floating(self, win, floating):
...
+                    self.tiled_windows.remove(win)
                 for i in self.layouts:
-                    i.remove(win)
-                    if win is self.current_window:
-                        i.blur()
+                    # Remove the window from the layout if it is not fullscreen
+                    if not win.fullscreen:
+                        i.remove(win)
+                        if win is self.current_window:
+                            i.blur()
```
[the remark this answers is not in the document]
Comment (jwijenbergh) [reply libqtile/group.py:303]: makes sense!

File: libqtile/group.py [hunk 4]
```diff
@@ -256,7 +259,8 @@ def remove(self, win, force=False):
                 or self.layout.focus_first()
                 or self.floating_layout.focus_first(group=self)
             )
-        else:
+        # Remove from the tiled layouts if it was not floating or in the tiled windows list
+        if not win.floating or win in self.tiled_windows:
```
[the remark this answers is not in the document]
Comment (jwijenbergh) [reply libqtile/group.py:263]: It is, done.

File: libqtile/group.py (hunk 1 above)
Comment (jwijenbergh) [reply libqtile/group.py:315]: Also makes sense, this is much better

File: libqtile/group.py (hunk 2 above)
Comment (jwijenbergh) [reply libqtile/group.py:49]: yep!

**Comment by jwijenbergh** [comment]
@m-col Thanks for the nice review. I have implemented all your suggestions. Could you give this a manual test? E.g. removing fullscreen windows, moving them to different groups and closing them, testing if they're always re-ordered correctly. This all worked for me but maybe you can find an edge case.

**Comment by m-col** [comment]
It seems to work perfectly for all layouts! It should be possible to actually test every layout in one go. In `test/layouts/test_common.py` check out `test_cycle_layouts`. Something similar could be done to check that the window order stays fixed even when cycling across layouts and toggling fullscreen each time.

**Comment by jwijenbergh** [comment]
> It seems to work perfectly for all layouts! It should be possible to actually test every layout in one go. In `test/layouts/test_common.py` check out `test_cycle_layouts`. Something similar could be done to check that the window order stays fixed even when cycling across layouts and toggling fullscreen each time.

Great idea. done!

**Comment by m-col** [comment]
Perfect, thanks! It will be auto-merged if the CI is happy :)

**Comment by jwijenbergh** [comment]
> Perfect, thanks! It will be auto-merged if the CI is happy :)

Thanks! I think you have to approve first though?

**Comment by m-col** [comment]
Looks like this caused a test failure on another PR: https://github.com/qtile/qtile/runs/4872898400?check_suite_focus=true#step:10:1337

I'll restart those tests and see if they are flaky or something.

**Comment by jwijenbergh** [comment]
> Looks like this caused a test failure on another PR: https://github.com/qtile/qtile/runs/4872898400?check_suite_focus=true#step:10:1337
> 
> I'll restart those tests and see if they are flaky or someth

> Looks like this caused a test failure on another PR: https://github.com/qtile/qtile/runs/4872898400?check_suite_focus=true#step:10:1337
> 
> I'll restart those tests and see if they are flaky or something.

Did restarting them work? I think it did right?

**Comment by m-col** [comment]
The test error happened again on master (following an unrelated PR that passed): https://github.com/qtile/qtile/runs/4914974040?check_suite_focus=true#step:10:1333

If you run it a bunch of times locally, maybe you can get it to fail?

**Comment by jwijenbergh** [comment]
> The test error happened again on master (following an unrelated PR that passed): https://github.com/qtile/qtile/runs/4914974040?check_suite_focus=true#step:10:1333
> 
> If you run it a bunch of times locally, maybe you can get it to fail?

I could try later but I'm not sure how this can be the case? There is no randomness involved.

**Comment by m-col** [comment]
Lol, no idea. Sometimes things that require clients to update their own state can be a bit racey, and those are often fixed by doing `manager.c.sync()`. However, I don't think that would be the case here. Perhaps debug logs would be needed... I'm not actually sure how to get those from tests.

**Comment by jwijenbergh** [comment]
> Lol, no idea. Sometimes things that require clients to update their own state can be a bit racey, and those are often fixed by doing `manager.c.sync()`. However, I don't think that would be the case here. Perhaps debug logs would be needed... I'm not actually sure how to get those from tests.

Sorry for the delay on this. I got to reproduce this locally (like 1-3 tests out of 300 sometimes) using (run in `test` directory):

```bash
#!/usr/bin/env bash
# Counter for how many times the test has failed
COUNTER=0

# Run tests 300 times
for i in {1..300}
do
    echo "RUNNING TEST $i"
    pytest -s test_group.py -k 'test_tiled_window_set' >logtemp 2>&1
    if [ $? -ne 0 ]; then
	((COUNTER=COUNTER+1))
	cat logtemp
    fi
done

echo "Failed $COUNTER times"
```

Not sure how to fix the test as of yet. I have asserted that the correct window is focused which it is, so it seems to be a problem only sometimes in `toggle_floating` during testing
